### ISSUE TITLE
Add generic boundaries to everyItem factory method

### ISSUE THREAD

**Issue by bjoernpollex** [hamcrest/JavaHamcrest#40]
The factory method `Every.everyItem` currently returns a `Matcher<Iterable<U>>`. Wouldn't a `Matcher<Iterable<? extends U>>` make sense here? This would be in accordance with the [guidelines for wildcard use](http://docs.oracle.com/javase/tutorial/java/generics/wildcardGuidelines.html).

Other collection matchers, such as `IsIterableContainingInOrder` already do this.

**Comment by fbastien** [comment]
@npryce Strongly disagree!!!
I'm unable to upgrade from 1.3 to 2.x because of this change, since my tests using `everyItem` wouldn't compile anymore. 

Actually @bjoernpollex misinterpreted the [guidelines for wildcard use](https://docs.oracle.com/javase/tutorial/java/generics/wildcardGuidelines.html). See in particular this note:

> These guidelines do not apply to a method's return type. Using a wildcard as a return type should be avoided because it forces programmers using the code to deal with wildcards.

Thus `everyItem` should keep returning `Matcher<Iterable<U>>`. A wildcard was indeed missing, but it should have been inserted on the item matcher parameter instead. Since the item matcher is in "out" mode (because it will receive items from the iterable), it must be written `Matcher<? super U>`.

To sum up, the factory signature should be `Matcher<Iterable<U>> everyItem(Matcher<? super U> itemMatcher)`, and the `Every` class should extend `TypeSafeDiagnosingMatcher<Iterable<T>>` (without wildcard).
(My tests compile again in my project after applying this change)

**Comment by nhojpatrick** [comment]
Just checking the current code and it's `public static <U> Matcher<Iterable<? extends U>> everyItem(final Matcher<U> itemMatcher) {`

And you and it to be; `public static <U> Matcher<Iterable<U>> everyItem(final Matcher<? extends U> itemMatcher) {`

**Comment by nhojpatrick** [comment]
Does this solution break https://github.com/hamcrest/JavaHamcrest/issues/289
